File: android/ccda-engine/src/main/kotlin/com/ccdakit/engine/Models/Media/CCDAMedia.kt
package com.ccdakit.engine

/**
 * Embedded or referenced media parsed from C-CDA `observationMedia` entries.
 *
 * @property id Stable media identity.
 * @property mediaType Typed media/MIME value.
 * @property representation C-CDA representation metadata.
 * @property payload Payload storage for the media.
 */
data class CCDAMedia(
    val id: String,
    val mediaType: CCDAMediaType,
    val representation: CCDAMediaRepresentation,
    val payload: CCDAMediaPayload,
) {
    /** Whether the media type represents an image. */
    val isImage: Boolean
        get() = mediaType.mimeType?.startsWith("image/") == true

    /** Whether this media payload is stored outside the model. */
    val isCached: Boolean
        get() = payload is CCDAMediaPayload.CachedFile

    /** Loads decoded media bytes from the payload. */
    fun loadData(): ByteArray = when (payload) {
        is CCDAMediaPayload.CachedFile -> payload.file.readBytes()
        is CCDAMediaPayload.InlineBase64 -> java.util.Base64.getDecoder()
            .decode(payload.value.filterNot { it.isWhitespace() })
        CCDAMediaPayload.Unavailable -> throw CCDAMediaPayloadException.Unavailable
    }
}

File: apple/Sources/CCDAEngine/Models/Media/CCDAMedia.swift
import Foundation

/// Embedded or referenced media parsed from C-CDA observationMedia entries.
public struct CCDAMedia: Identifiable, Sendable {
    /// Media identifier from XML or a generated fallback.
    public let id: String
    /// Typed media/MIME value.
    public let mediaType: CCDAMediaType
    /// C-CDA representation value.
    public let representation: CCDAMediaRepresentation
    /// Payload storage for the media.
    public let payload: CCDAMediaPayload

    /// Loads decoded media data from the payload.
    public func loadData() throws -> Data {
        switch payload {
        case .cachedFile(let url, _):
            return try Data(contentsOf: url)
        case .inlineBase64(let value):
            guard let data = Data(base64Encoded: value.filter { !$0.isWhitespace }) else {
                throw CCDAMediaPayloadError.invalidBase64
            }
            return data
        case .unavailable:
            throw CCDAMediaPayloadError.unavailable
        }
    }

    /// Whether the media type represents an image.
    public var isImage: Bool {
        mediaType.isImage
    }

    /// Whether this media payload is stored outside the model.
    public var isCached: Bool {
        if case .cachedFile = payload {
            return true
        }
        return false
    }

    /// Creates a media value.
    public init(
        id: String,
        mediaType: CCDAMediaType,
        representation: CCDAMediaRepresentation,
        payload: CCDAMediaPayload
    ) {
        self.id = id
        self.mediaType = mediaType
        self.representation = representation
        self.payload = payload
    }
}
